### PR TITLE
Backport #16550 to main: Fix bug for integer compressed images with blanks

### LINKED ISSUE #15236: New FITS compression code breaks loading Pan-STARRS images
### Description

The code for reading compressed FITS files in 5.3 is crashing on loading the images from Pan-STARRS. It was working nicely in 5.2.

The problem manifests like that:
```
$ python -c 'from astropy.io import fits; fits.getdata("http://ps1images.stsci.edu/rings.v3.skycell/1004/016/rings.v3.skycell.1004.016.stk.g.unconv.fits")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/ltzfs/ASTRO101/anaconda3/envs/astro101/lib/python3.11/site-packages/astropy/io/fits/convenience.py", line 232, in getdata
    data = hdu.data
           ^^^^^^^^
  File "/ltzfs/ASTRO101/anaconda3/envs/astro101/lib/python3.11/site-packages/astropy/utils/decorators.py", line 837, in __get__
    val = self.fget(obj)
          ^^^^^^^^^^^^^^
  File "/ltzfs/ASTRO101/anaconda3/envs/astro101/lib/python3.11/site-packages/astropy/io/fits/hdu/compressed.py", line 1558, in data
    data = self.section[...]
           ~~~~~~~~~~~~^^^^^
  File "/ltzfs/ASTRO101/anaconda3/envs/astro101/lib/python3.11/site-packages/astropy/io/fits/hdu/compressed.py", line 2189, in __getitem__
    data = decompress_hdu_section(self.hdu, first_tile_index, last_tile_index)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ltzfs/ASTRO101/anaconda3/envs/astro101/lib/python3.11/site-packages/astropy/io/fits/_tiled_compression/tiled_compression.py", line 442, in decompress_hdu_section
    tile_data[blank_mask] = np.nan
    ~~~~~~~~~^^^^^^^^^^^^
ValueError: cannot convert float NaN to integer
```

### Expected behavior

The images from Pan-STARRS were loading without issues in pre-5.3

### How to Reproduce

```python
from astropy.io import fits
fits.getdata("http://ps1images.stsci.edu/rings.v3.skycell/1004/016/rings.v3.skycell.1004.016.stk.g.unconv.fits")
```

### Versions

Linux-4.18.0-372.9.1.el8.x86_64-x86_64-with-glibc2.28
Python 3.11.4 (main, Jul  5 2023, 13:45:01) [GCC 11.2.0]
astropy 5.3.2
Numpy 1.25.2
pyerfa 2.0.0.3
Scipy 1.11.2
Matplotlib 3.7.2

### LINKED ISSUE #16631: Reminder: port regression tests from #16550 to astropy 7.0
This is a reminder from https://github.com/astropy/astropy/pull/16550#issuecomment-2189916395 that tests from #16550 should be forward ported to the dev branch (7.0) after #15474 is merged.
@astrofrog

### PR DESCRIPTION
**Pull request by saimn**
Fix #15236
Fix #16631

### Description

This pull request is to address ...

Fixes #<Issue Number>

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Maybe we should credit @rlwastro as the co-author here since this is a manual forward-port? (Huh, why did we put the fix in backport branch only if forward port is needed anyway?)

**Comment by saimn** [comment]
Good point, I added @rlwastro as coauthor. The fix was made on v6.1.x in order to have it released quickly, without risking conflict with the refactor PR. Now that the refactor is done I could simplify a bit the fix btw.

**Comment by pllim** [comment]
@saimn , do you want to squash 4 of the 5 commits? Otherwise, I'll have to use the "squash and merge" button to keep things clean.

**Comment by neutrinoceros** [comment]
This also fixes #16631 !

File: astropy/io/fits/hdu/compressed/tests/test_tiled_compression.py [hunk 1]
```diff
@@ -65,6 +66,97 @@ def test_zblank_support(canonical_data_base_path, tmp_path):
...
+    cfile1 = tmp_path / "compressed_with_blank.fits"
+    hdu = fits.CompImageHDU(
+        data=data, header=header, compression_type="RICE_1", tile_shape=(16, 16)
+    )
+    hdu.writeto(cfile1, overwrite=True, checksum=False)
+
+    # replace BLANK keyword in header with ZBLANK keyword
+    cfile2 = tmp_path / "compressed_with_zblank.fits"
+    with fits.open(cfile1, disable_image_compression=True) as hdul:
+        assert "ZBLANK" not in hdul[1].header
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/io/fits/hdu/compressed/tests/test_tiled_compression.py:104]: Ah wait ignore me, I just read the FITS standard :)

File: astropy/io/fits/hdu/compressed/_tiled_compression.py [hunk 2]
```diff
@@ -359,7 +360,7 @@ def decompress_image_data_section(
     else:
         zzero_column = None
 
-    zblank_header = compressed_header.get("ZBLANK", None)
+    zblank_header = compressed_header.get("ZBLANK", image_header.get("BLANK", None))
```
Comment (astrofrog) [inline astropy/io/fits/hdu/compressed/_tiled_compression.py:363]: Just to check, is BLANK not defined also in the compressed header if we are decompressing data? Or is the idea here to allow a user to change BLANK after opening a file but before decompressing it?

**Review by astrofrog — APPROVED** [review]
This seems ok to me, took a little while to properly understand!

File: astropy/io/fits/hdu/compressed/_tiled_compression.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/hdu/compressed/_tiled_compression.py:363]: It doesn't work for the last test case: zblank column without z/blank in the header. In that case we use the blank value that is set below in `header.py` and is not in the compressed header.
I tried to remove that line, we could pass the blank value by another mean, but in the end it was not looking much better so I kept that.

**Comment by saimn** [comment]
Ok let's merge this one so we close a few more tickets :)

**Pull request merged**